### ISSUE TITLE
Prototype CUDA graphs user exposure for 1.10

### ISSUE THREAD

**Issue by mcarilli** [pytorch/pytorch#61880]
## 🚀 Feature
I plan to add and document _prototype_ user-facing Python exposures for existing CUDA graphs bindings.

PR in progress: https://github.com/pytorch/pytorch/pull/63269

@mruberry and @ngimel want this for 1.10. IMO that's enough time as long as we're comfortable calling it a prototype.

## Motivation

Raw CUDA graphs Python bindings already exist:
```python
g = torch.cuda._Graph()
g.capture_begin()
g.capture_end()
g.replay()
```
but there's no documented user-facing API or examples.

## Pitch

* Wrap the raw bindings to provide two capture options:

1. a context manager to capture arbitrary ops, ideal if the entire network is capturable:
```python
# user must run a few warmup iterations first, e.g. to get cudnn benchmarking out of the way
g = torch.cuda._Graph()
with torch.cuda.graph(g):
    static_loss = net(static_in)
    static_loss.backward()
    optimizer.step()

for d in data:
    static_in.copy(d)
    g.replay() # plays the whole fwd, bwd, step. Efficient!
    # you don't even need zero_grad because replay() refills the static .grad attributes in place.
```
2. a helper to create an autograd-aware graphed section of network, necessary if only part of a network is capturable:
```python
# Takes care of warmup iterations,
# sets up graphed forward and backward passes for the submodule  and wraps them in the forward and backward methods of a torch.autograd.Function,
# replaces graphsafe_module.forward with the Function's .apply.
net.graphsafe_module = torch.cuda.make_graphed_module**(net.graphsafe_module, sample_arg)

# training loop looks typical
for _ in range(10):
    optimizer.zero_grad(set_to_none=True)
    loss = net(x) # net.graphsafe_module executes graphed forward, other submodules run eagerly
                  # graphsafe_module also internally maintains the graph's static input buffers, and copies
                  # new inputs to those buffers (unlike 1., no manual static_in.copy_ is needed)
    loss.backward() # graphsafe_module's bwd kernels run in a graph, bwd ops from other submodules run eagerly
    optimizer.step()
    
** working title
```

* Document 1. and 2. under torch.cuda.

* Add examples to the CUDA Semantics note.

## Alternatives

The API will be subject to change. It's hard to anticipate everything people need for complicated networks, but we have to start somewhere.

Error-checking in the above APIs is uncertain (my local implementations rely a lot on user judgment) and will evolve based on what gotchas turn out to be common.

@ezyang @ptrblck @csarofeen @gchanan

**Comment by byronyi** [comment]
```
with torch.cuda.graph(g):
    static_loss = net(static_in)
    static_loss.backward()
    optimizer.step()
```

Does this support dynamic loss scaling optimizer? We observe a similar problem on torch-xla side which uses graph capture as well: https://github.com/pytorch/xla/issues/3043#issuecomment-890285855

**Comment by mcarilli** [comment]
@byronyi You're absolutely right. The recommended approach is simple: don't capture scaler.step(optimizer). I included an example in the PR: https://github.com/pytorch/pytorch/pull/63269/files#diff-4840823ff4b0cd7fe61b9ada2a1746f6936c944bfbcc4d1bdb8b3dc1561d4401R773-R816

**Comment by mcarilli** [comment]
>The long term solution is probably to wrap the dynamic loss scaler in torch.jit.script and lower it to a conditional node in graph IR.

Sorry, the word "graph" is used for a million things these days.
Cuda graph capture is a cuda-internal thing. Cuda itself just records whatever kernels run during capture (with args). It has no awareness of `torch.jit.script`'s "graph" or IR.

Hypothetically, pytorch jit could rig some communication with capture: for example, a jit conditional node could detect if capture is underway, and if so, call [cudaStreamGetCaptureInfo_v2](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g0a740cbac128537e59bc6307a2b301b5) on the current stream to get a handle to graph currently being constructed, then use the [manual graph-building API](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#creating-a-graph-using-api) to insert a cuda graph conditional node at that point in the capturing stream. But I don't know if "cuda graph conditional nodes" are even available in the current release of cuda. Overall, this feature would be cool, but technically challenging, and today just a dream.

For dynamic loss scaling optimizers in particular it's not worth the complexity, since not capturing scaler.step(optimizer) is an easy workaround.

Admittedly, not capturing scaler.step(optimizer) is not ideal. Optimizers often have loops over long parameter lists that end up CPU bound, which means they could benefit from graph capture.

https://github.com/pytorch/pytorch/tree/master/torch/optim/_multi_tensor optimizers are a project underway to greatly reduce optimizer CPU overhead without graphs, by updating many tensors at a time in just one or a few cuda kernel launches.

You could also consider graphing `optimizer.step()` by itself, then patching `optimizer.step` with a one-line function that runs the graph. Later, when you run `scaler.step(optimizer)` _ungraphed_ in the training loop, the inf/nan check will take place _ungraphed_, but when `scaler.step` chooses to run `optimizer.step()` internally, your patched `optimizer.step` runs optimizer math as a graph.

**Comment by mcarilli** [comment]
Closing, as https://github.com/pytorch/pytorch/pull/63269 was merged in time for 1.10.

**Comment by byronyi** [comment]
See https://github.com/pytorch/xla/issues/3120 for a clever trick to address the GradScalar issue while there is no host-device sync involved.

**Comment by mcarilli** [comment]
Interesting. I don't understand why their trick helps. Replied on that thread https://github.com/pytorch/xla/issues/3120#issuecomment-917699833

**Comment by ekhahniii** [comment]
Hi @mcarilli, super excited for this work. Is there a good place to provide feedback on the beta API and track upcoming plans for this feature?

**Comment by ngimel** [comment]
YOu can either leave feedback here, or open a new issue and label it with "module: cuda graphs"
